### PR TITLE
Apply rates to chain item drops

### PR DESCRIPTION
**Pull request by olsonpm**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
apply `battle_config.item_rate_adddrop` to rates defined in `item_chain.conf`

**Issues addressed:** #2982 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/itemdb.c [hunk 1]
```diff
@@ -1571,7 +1571,12 @@ static void itemdb_read_chains(void)
 				prev->next = &itemdb->chains[count].items[c - 1];
 
 			itemdb->chains[count].items[c - 1].id = data ? data->nameid : 0;
-			itemdb->chains[count].items[c - 1].rate = data ? libconfig->setting_get_int(entry) : 0;
+
+			int rate = data ? libconfig->setting_get_int(entry) : 0;
+			if (battle_config.item_rate_adddrop != 100)
+				itemdb->chains[count].items[c - 1].rate = rate*battle_config.item_rate_adddrop/100;
```
Comment (4144) [inline src/map/itemdb.c:1577]: missing spaces near ``*`` and near ``/``

**Review by 4144 — CHANGES_REQUESTED** [review]

File: src/map/itemdb.c (hunk 1 above)
Comment (olsonpm) [reply src/map/itemdb.c:1577]: so I copied that [from here](https://github.com/HerculesWS/Hercules/blob/fbd9c8f9df8022619dcae158664bb5e81c7ae4bf/src/map/pc.c#L2163)

```c
rate = rate*battle_config.item_rate_adddrop/100;
```

is there a linter/formatter/other tool I can use to catch these ?

Comment (4144) [reply src/map/itemdb.c:1577]: many code in hercules with wrong style. if if line changed or added in commit it should use correct style
don't need search other style issues. only follow style in new/changed code

Comment (olsonpm) [reply src/map/itemdb.c:1577]: thanks that makes sense.  for future prs though, are there any tools which will catch style issues like this?  like linters or formatters?

Comment (4144) [reply src/map/itemdb.c:1577]: no, only manual checking

**Comment by dastgirp** [comment]
I Strongly suggest to add new config, as this would break custom changes for many servers whenever they update it.

There isn't any warning or info displayed to them that this flag now affects other drops as well.

File: src/map/itemdb.c (hunk 1 above)
Comment (olsonpm) [reply src/map/itemdb.c:1577]: okay no problem, i'll make sure to check next time

**Comment by olsonpm** [comment]
@dastgirp i am willing to change the pr to affect only cards, not all chain items, but as far as a new config is concerned - this is a bug fix as the current config says it affects all card drops should be affected by the rate.  and currently cards which give drops via item cain groups are not impacted

**Comment by dastgirp** [comment]
> @dastgirp i am willing to change the pr to affect only cards, not all chain items, but as far as a new config is concerned - this is a bug fix as the current config says it affects all card drops should be affected by the rate.  and currently cards which give drops via item cain groups are not impacted

it says card-granted item drops, and not card drops.
That means bonuses which add additional drops.

Commit in which it was implemented: https://github.com/HerculesWS/Hercules/commit/ee8093a79840edc7169291ca69ac3d256190fc83 (commit message/readme have much more clear description)

**Comment by olsonpm** [comment]
card granted item drops don't include item chain?  sorry i'm not in front of my computer, i can check the commit message another time

**Comment by olsonpm** [comment]
brazer card is a card which applies an item chain drop.  i feel this config should cover that card

**Comment by olsonpm** [comment]
Okay I read the commit message and changelog entry.  I still don't understand why the config would omit rates defined in the item-chain when they are dropped via cards.  That feels like a bug to me.

**Comment by 4144** [comment]
after your changes server look like crashing

About config option i think new option was requested because if bug exists for too long it converted to feature. I mean all servers know this bug exists and using it as is.

But i not sure is really need another option or not. may be ask other devs?

**Comment by olsonpm** [comment]
Yeah I'll take a look at the pr errors when I'm back at my comp.

And a new config option makes sense considering the bug has existed for a long time.  I can do that

**Comment by olsonpm** [comment]
So the error is caused by my lack of c knowledge - I didn't realize structs were copy when assigned (i.e. pass by value).

Is there a recommended way to avoid the boilerplate of `itemdb->chains[count].items[c - 1]` everywhere ?  I would like to use a variable for this but the only thing I can come up with is storing a pointer to the struct e.g.

```c
struct item_chain_entry *item = &itemdb->chains[count].items[c - 1];

if( prev )
  prev->next = item;

(*item).id = data ? data->nameid : 0;

int rate = data ? libconfig->setting_get_int(entry) : 0;
if (battle_config.item_rate_adddrop != 100)
  (*item).rate = rate * battle_config.item_rate_adddrop / 100;
else
  (*item).rate = rate;
// ...
```

that feels weird, but I also am not a c programmer so I don't know what's preferred

**Comment by olsonpm** [comment]
@4144 @dastgirp the new config option was implemented.  If there are build errors I may not have time to get to them til monday/tuesday

**Review by 4144 — APPROVED** [review]

**Comment by 4144** [comment]
probably better squash commits into one or into two

**Comment by olsonpm** [comment]
sure thing

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**